runtime/lua

justinmk opened the pull request:
Followup to baby steps in https://github.com/neovim/neovim/pull/9280 .

Related: #9292
Related: #6580

- lua/stdlib: Introduce `vim.shared` module
  - This is where "pure functions" can live, which can be shared by Nvim and test logic which may not have a running Nvim instance available.
    - If in the future we use Nvim itself as the Lua engine for tests, then these functions could be moved directly onto the `vim` Lua module.
- Test `helpers` modules are all combined. E.g. `require('test.functional.helpers')` includes all of the public methods from `require('test.helpers')`. No need for `global_helpers = require('test.helpers')`.

justinmk committed:
test: Extend {unit,functional}.helpers with global helpers

Automatically include all "global helper" util functions in the
unit.helpers and functional.helpers and modules.  So tests don't need to
expicitly do:

    local global_helpers = require('test.helpers')

justinmk committed:
lua/stdlib: Introduce vim.shared

This is where "pure functions" can live, which can be shared by Nvim and
test logic which may not have a running Nvim instance available.

If in the future we use Nvim itself as the Lua engine for tests, then
these functions could be moved directly onto the `vim` Lua module.

closes #6580

mhinz commented on test/helpers.lua:
```diff
@@ -428,6 +430,34 @@ local function concat_tables(...)
+    if tbl then
+      for k, v in pairs(tbl) do
+        if behavior ~= 'force' and ret[k] ~= nil then
+          if behavior == 'error' then
```
`behaviour == 'error'` although this branch is only taken if `behaviour ~= 'force'`? Is this ever true?

justinmk replied:
why wouldn't it be? `'force` ~= 'error'`

justinmk committed:
test: share implementation of testdir/load.vim

Also, don't compute load factor unless load_adjust() was called, it
slows down the test suite.

ref #9292

justinmk commented on test/functional/helpers.lua:
```diff
@@ -747,41 +739,14 @@ local function alter_slashes(obj)
-end
-
--- Compute load factor only once.
-local load_factor = compute_load_factor()
```
@mhinz this was running for all tests, even those that don't need `load_adjust()`.  I guess that could slow down the tests.

justinmk committed:
test: remove use of require('test.helpers')

The test.functional.helpers and test.unit.helpers modules now include
all of the public functions from test.helpers, so there is no need to
separately require('test.helpers').

justinmk committed:
lua/shared: move table util funcs to vim.shared

Use `tbl_` prefix for all table-util functions. Specify in the function
docstring if it expects a list-like or map-like table.
